### PR TITLE
projection validation test passed

### PR DESCRIPTION
**Pull request by angsherpa232**

### CODE REVIEW

File: src/project.js [hunk 1]
```diff
@@ -32,7 +32,13 @@ const conditions = [
   },
 ];
 
-function project(location) {
+const height = types.heightLevel
```
Comment (friedjoff) [inline src/project.js]: Better name this `heightLevelList` directly.

File: src/project.js [hunk 2]
```diff
@@ -32,7 +32,13 @@ const conditions = [
...
+  .map(h => h.key)
+  .splice(0, 7)
```
Comment (friedjoff) [inline src/project.js]: Why do you splice?

File: src/project.js [hunk 3]
```diff
@@ -32,7 +32,13 @@ const conditions = [
...
+  .reverse();
+
+let initIterValue = 0;
+function project(location, targetHeight) {
```
Comment (friedjoff) [inline src/project.js]: Let's try to use consistent names, so better `targetHeightLevel`.

File: src/project.js [hunk 4]
```diff
@@ -57,7 +63,22 @@ function project(location) {
   if (typeof target === 'string') {
     newLocation.target = target;
   }
-  return newLocation;
+
+  const heightLevelList = height;
+  initIterValue += 1;
+  let result = newLocation;
```
Comment (friedjoff) [inline src/project.js]: Do we need result as a intermediate variable. Can't we use newLocation directly?

File: src/project.test.js [hunk 5]
```diff
@@ -2,23 +2,34 @@ const { project } = require('../');
...
+    ).forestType,
+  ).toBe('1');
 });
 
 test('invalid location values', () => {
   expect(() =>
     project({ forestType: '60*', forestEcoregion: 'fooBar' }),
   ).toThrowError('fooBar for forestEcoregion is not valid.');
 
+  // expect(() =>
```
Comment (friedjoff) [inline src/project.test.js]: Please only add running tests.

File: src/project.js [hunk 6]
```diff
@@ -57,7 +63,22 @@ function project(location) {
   if (typeof target === 'string') {
     newLocation.target = target;
   }
-  return newLocation;
+
+  const heightLevelList = height;
+  initIterValue += 1;
```
Comment (friedjoff) [inline src/project.js]: What about naming this `heightLevelPosition`?

File: src/project.test.js [hunk 7]
```diff
@@ -2,23 +2,34 @@ const { project } = require('../');
 
 test('valid projection', () => {
   expect(
-    project({
```
Comment (friedjoff) [inline src/project.test.js]: Better keep this test for a valid single-step projection and add a new test for multi-step projections.

File: src/project.js (hunk 2 above)
Comment (angsherpa232) [reply src/project.js]: beacuse splicing make sures that the first heightlevel is Obersubalpin. Otherwise the list would start from unter-/obermontan and recursion would happen 4 more times which is unnecessary.

File: src/project.js (hunk 4 above)
Comment (angsherpa232) [reply src/project.js]: Much better.

File: src/project.test.js (hunk 5 above)
Comment (angsherpa232) [reply src/project.test.js]: removed the commented one. All other tests are running.

File: src/project.js (hunk 6 above)
Comment (angsherpa232) [reply src/project.js]: sounds good.

File: src/project.test.js (hunk 7 above)
Comment (angsherpa232) [reply src/project.test.js]: agree.

File: src/project.js (hunk 2 above)
Comment (friedjoff) [reply src/project.js]: Ok, understand. To me it looks like our heightlevel metadata is not optimal. Let's keep it like this, because I guess we need to refactor this a bit based on the new spatial data anyway.

File: src/project.js [hunk 8]
```diff
@@ -44,18 +52,37 @@ function project(location) {
...
       break;
     }
   }
-  const newLocation = { ...location, options };
-  if (typeof target === 'string') {
-    newLocation.target = target;
+
+  if (typeof forestType !== 'string') {
+    throw new Error(
+      'No height Level found. Try with another Target height Level.',
```
Comment (friedjoff) [inline src/project.js]: Is this a good error message? I guess this error is more general, more like "No projection found.". Please also add a test for this error.

Comment (angsherpa232) [reply src/project.js]: Yeah, good question. Modified it, now it should make more sense. Also added the test for this case. :smile:

File: data/projections.csv [hunk 9]
```diff
@@ -5,11 +5,11 @@ R,J,M,1,2,3 osa bis co;obersubalpin;99999;99999;99999;99999;;;;;;subalpin
 R,J,M,1,2,3 osa bis co;obersubalpin;59;< als 60% ;Boden normal;57V;;;;;;subalpin
 R,J,M,1,2,3 osa bis co;obersubalpin;59;< als 60% ;Boden normal;57V;;;;;;subalpin
 R,J,M,1,2,3 osa bis co;obersubalpin;99999;99999;99999;99999;;;;;;subalpin
-R,J,M,1,2,3 osa bis co;obersubalpin;59Lä;> als 60% ;Boden normal;57CLä;;;;;;subalpin
-R,J,M,1,2,3 osa bis co;obersubalpin;59Lä;> als 60% ;Boden normal;57CLä;;;;;;subalpin
+R,J,M,1,2,3 osa bis co;obersubalpin;59LÃĊ;> als 60% ;Boden normal;57CLÃĊ;;;;;;subalpin
```
Comment (friedjoff) [inline data/projections.csv]: Seems like this introduce a encoding bug. Can you please check why this changes?

File: data/projections.json [hunk 10]
```diff
@@ -1 +1,6560 @@
...: {"unknown": {"unknown": {"unknown": "23H"}}}}}, "2a": {"C": {"unknown": {"unknown": {"unknown": {"unknown": "23H"}}}}}, "2b": {"C": {"unknown": {"unknown": {"unknown": {"unknown": "23H"}}}}}}, "25F": {"1": {"SM": {"unknown": {"unknown": {"unknown": {"unknown": "25F"}}}}, "UM": {"unknown": {"unknown": {"unknown": {"unknown": "25F"}}}}}, "3": {"SM": {"unknown": {"unknown": {"unknown": {"unknown": "24*"}}}}, "UM": {"unknown": {"unknown": {"unknown": {"unknown": "25F"}}}}}, "J": {"SM": {"unknown": {"unknown": {"unknown": {"unknown": "25F"}}}}, "UM": {"unknown": {"unknown": {"unknown": {"unknown": "25F"}}}}}, "M": {"SM": {"unknown": {"unknown": {"unknown": {"unknown": "25F"}}}}, "UM": {"unknown": {"unknown": {"unknown": {"unknown": "25F"}}}}}, "2a": {"SM": {"unknown": {"unknown": {"unknown": {"unknown": "25F"}}}}, "UM": {"unknown": {"unknown": {"unknown": {"unknown": "25F"}}}}}, "2b": {"SM": {"unknown": {"unknown": {"unknown": {"unknown": "24*"}}}}, "UM": {"unknown": {"unknown": {"unknown": {"unknown": "25F"}}}}}}, "25Q": {"1": {"C": {"unknown": {"unknown": {"unknown": {"unknown": "52"}}}}}, "J": {"C": {"unknown": {"unknown": {"unknown": {"unknown": "52"}}}}}, "M": {"C": {"unknown": {
```
Comment (friedjoff) [inline data/projections.json]: Please only add unformatted JSON files.

File: src/project.js [hunk 11]
```diff
@@ -44,18 +52,38 @@ function project(location) {
...
       break;
     }
   }
-  const newLocation = { ...location, options };
-  if (typeof target === 'string') {
-    newLocation.target = target;
+
+  if (typeof forestType !== 'string') {
+    throw new Error(
+      'primitive datatype is not valid for projected forestType.',
```
Comment (friedjoff) [inline src/project.js]: This error message is still very technical and referring to internal implementation details. Something like "Found no projection for location." seems more appropriate. Also a test to catch this error is still missing.

Comment (angsherpa232) [reply src/project.js]: Totally, made something like "Found no projection for selected target height.". The test should be in the test file, the third one.

File: data/projections.json (hunk 10 above)
Comment (angsherpa232) [reply data/projections.json]: Got it.

File: data/projections.csv (hunk 9 above)
Comment (angsherpa232) [reply data/projections.csv]: Found the reason, seems like it happened during the encoding to UTF-8. The original file is safe so will fix it soon.

File: src/project.test.js [hunk 12]
```diff
@@ -2,32 +2,81 @@ const { project } = require('../');
...
+        slope: 'unknown',
+        tannenareal: 'unknown',
+        relief: 'unknown',
+      },
+      'OM',
+    ).forestType,
+  ).toBe('1h');
+});
+
+test('valid data type for projection', () => {
```
Comment (friedjoff) [inline src/project.test.js]: This case is already covered by the first test. Instead we should replace this one with a test which catches 'Found no projection for selected target height.' (use `toThrowError`).
